### ISSUE TITLE
VirtualThreads cannot be used with JAX-RS interfaces

### ISSUE THREAD

**Issue by danielbobbert** [quarkusio/quarkus#45186]
### Describe the bug

We are implementing a JAX-RS endpoint, where the API is given by a Java interface (provided by another project). Our implementation consists of an ApplicationScoped bean implementing this interface. We would like to have the implementation execute on virtual threads, but Quarkus fails to recognize the annotation.

### Expected behavior

We should be able to put \@RunOnVirtualThread on the class or methods of our bean that implements the JAX-RS interface, to have them execute on a virtual thread.

### Actual behavior

Putting \@RunOnVirtualThread
- on the interface itself: ignored
- on a method of the interface: ignored
- on the implementing bean itself: ignored
- on a method of the implementing bean: build time error because the annotation is only allowed on "entrypoint methods"

There seems to be no valid way to get one of the methods to execute on a virtual thread.
This behaviour is present in Quarkus 3.17.4 but was already present in 3.8 (with quarkus-resteasy-reactive), so it seems to be general issue with the recognition of "entrypoints" in conjunction with Java interfaces.

### How to Reproduce?

I have attached a small reproducer tha shows that \@RunOnVirtualThread works correctly when put on a JAX-RS bean, but it is being ignored (or leads to an error message) when used in conjunction with interfaces.

### Output of `uname -a` or `ver`

Mac OS

### Output of `java -version`

Java 21

### Quarkus version or git rev

3.17.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvn 3.9.9

### Additional information

[virtual-rest-impl.zip](https://github.com/user-attachments/files/18180635/virtual-rest-impl.zip)

**Comment by geoand** [comment]
Thanks for reporting.

I am pretty sure we've discussed this before. Do you remember @michalvavrik @cescoffier @Ladicek ?

**Comment by cescoffier** [comment]
Yes, their was discussed. If I remember correctly, we said we did not want to support @RunOnVirtualThread on interfaces.

**Comment by cescoffier** [comment]
Actually the report is slightly different. Not sure how it can be fixed...

**Comment by danielbobbert** [comment]
IMHO, the annotation doesn't belong on the interface (as it is an implementation detail). What I am asking for is that the annotation be allowed and honoured on the implementing bean.
Any way to set a default would also help (akin to "quarkus.messaging.blocking.signatures.execution.mode" which allows to set a default for message consumers).
One can already annotate a JAX-RS "Application" class with \@Blocking or \@NonBlocking. Allowing \@RunOnVirtualThread there to set the "BlockingDefault" on the "ApplicationScanningResult" would also be an option.

I am aware of the risks of setting virtual threads as the default (that's why I am asking for annoating the implementing bean). But right now, there seems to be no way at all. So setting a default (where it is known to be safe) would still be better than nothing...

**Comment by danielbobbert** [comment]
I guess, the proper "fix" would be to check implemented interfaces for JAX-RS annotations when collecting "entrypoint" methods. Not sure how complicated that could become, and maybe it's OK to prohibit multiple inheritance and such. But the plain case of "Bean implements Intfce" where the JAX-RS annotations are on the interface would be really helpful (and not an uncommon case for "API-first" development cases).

**Comment by geoand** [comment]
I'll have a look

**Comment by geoand** [comment]
Actually cc @Ladicek who did the implementation of `ExecutionModelAnnotationsProcessor`

**Comment by danielbobbert** [comment]
Well, the `ExecutionModelAnnotationsProcessor` only covers the "symptoms". Even if I disable the checking by setting the "detection-mode" to "disabled" I cannot get the methods to execute on a virtual thread, no matter where I put the annotation (class vs method, bean vs interface). So I think the problem is rather in the area of detecting endpoint methods. I wonder whether `org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod()` would be a good point to look at?!

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java#L844 is the place

**Comment by danielbobbert** [comment]
I think, I found one (other?) part of the problem, while debugging through EndpointIndexer.isRunOnVirtualThread().
This is called twice from createResoucreMethod(): once for the "currentMethodInfo" and once for the "actualMethodInfo" (which differ in the case of "bean implements interface").
Now in the first case, the "default" passed to isRunOnVirtualThread() is "AUTOMATIC". But in the second case (actualMethodInfo), "BLOCKING" is passed (see explicit comment in line 731). But the way isRunOnVirtualThread() is written, this default actually overwrites the explicit recognition of the annotation.

**Comment by danielbobbert** [comment]
In my testing, when passing AUTOMATIC to the second (actualMethodInfo) invocation of `isRunOnVirtualThread()`, then the annotation seems to be honoured correctly on the implementing bean (both on class and method level).
To be honest, I don't really understand the comment in line 738. Is this supposed to be a performance optimization? I can't spot anything in `isBlocking()` and `isRunOnVirtualThread()` that would make a difference though (besides maybe `doesMethodHaveBlockingSignature()`being called at the end of `isBlocking()`.

**Comment by geoand** [comment]
> Or in short: "explicit annotation beats default"

This is indeed the expected behavior.

**Comment by danielbobbert** [comment]
Will you be fixing this or do you want me to create a PR?

**Comment by geoand** [comment]
I'm busy with other stuff ATM, so if you have a fix in mind, feel free to open a PR
